src/parser/MagicServiceQuery.h

```diff
@@ -0,0 +1,69 @@
+// Copyright 2024, University of Freiburg
+// Chair of Algorithms and Data Structures
+// Authors: Johannes Herrmann <johannes.r.herrmann(at)gmail.com>
+//          Christoph Ullinger <ullingec@informatik.uni-freiburg.de>
+
+#pragma once
+
+class SparqlTriple;
+
+namespace parsedQuery {
+
+struct BasicGraphPattern;
+
+class MagicServiceException : public std::exception {
+  std::string message_;
+
+ public:
+  explicit MagicServiceException(const std::string& message)
+      : message_(message) {}
+  const char* what() const noexcept override { return message_.data(); }
+};
+
+// Abstract structure for parsing a magic SERVICE statement (used to invoke
+// QLever-specific features)
```

Comment:
Also document the structure of the base class (in particular you expect the configuration to consist only of triples, (which are handled by the subclasses via trhe virtual `addParameter` function, and (optionally) a single GroupGraphPattern, which is handled by the base class.